[7102][fb_bootcamp] Moved Gesv to ATen

animesht opened the pull request:
#7102

Moved _gesv_single and _gesv_single_out to ATen

animesht committed:
Moved Gesv to ATen

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -85,14 +112,14 @@ std::tuple<Tensor,Tensor> _gesv_helper_cpu(const Tensor& self, const Tensor& A)
+  if (!batched) {
+    Tensor A_;
+    Tensor b_;
+    return _gesv_single_helper(self, A, b_, A_);
```
this doesn't work for cuda.

ssnl reviewed (dismissed):
`_gesv_single_helper` doesn't work for cuda. `_th_gesv_single` isn't removed.

animesht committed:
Removed TH code, added CUDA support

SsnL commented on aten/src/ATen/native/cuda/Gesv.cu:
```diff
@@ -138,6 +138,36 @@ std::tuple<Tensor,Tensor> _gesv_helper_cuda(const Tensor& self, const Tensor& A)
+      ALLOCATE_ARRAY(info_array, magma_int_t, 1, result0);
+      ALLOCATE_ARRAY(ipiv_array, magma_int_t*, 1, result0);
+
+      magmaGesvBatched<scalar_t>(
```
for single gesv on cuda, you should use `magma_sgesv_gpu` or `magma_dgesv_gpu`. see https://github.com/pytorch/pytorch/blob/master/aten/src/THC/generic/THCTensorMathMagma.cu#L63-L99

ssnl reviewed (commented):
need to remove thc gesv code too.

SsnL commented:
test failure still looks legit

animesht committed:
Fixed CUDA support, removed gesv from THC, fixed autograd+tests

SsnL commented:
@pytorchbot retest this please

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +72,40 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+  AT_DISPATCH_FLOATING_TYPES(self.type(), "gesv", [&]{
+    auto A_ptr = result1.data<scalar_t>();
+    auto b_ptr = result0.data<scalar_t>();
+    auto ipiv = at::empty({bx}, result0.type().toScalarType(kInt));
```
use `result0.options().dtype(kInt)`.

SsnL commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -720,6 +720,17 @@
     CPU: _gesv_helper_cpu
     CUDA: _gesv_helper_cuda
 
+- func: _gesv_single(Tensor self, Tensor A) -> (Tensor, Tensor)
```
this doesn't need two implementations, right?

SsnL replied:
I meant that `_gesv_single_cpu` and `_gesv_single_cuda` can be merged together. They only differ by the `_gesv_single_out_cpu` vs. `_gesv_single_out_cuda` call, but you can use `_gesv_single_out` to dispatch to correct version.

animesht committed:
Used options().dtype(kInt)

animesht committed:
addressed comments - simplified gesv_single

animesht committed:
Fixed autograd

animesht commented:
@pytorchbot retest this please

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -114,13 +146,14 @@ std::tuple<Tensor,Tensor> gesv(const Tensor& self, const Tensor& A) {
 
 std::tuple<Tensor&,Tensor&> gesv_out(
-    Tensor& solution, Tensor& lu, const Tensor& self, const Tensor& A) {
+    Tensor& result0, Tensor& result1, const Tensor& self, const Tensor& A) {
```
hmm why did you change the variable names? they shouldn't matter, right?

SsnL commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -711,7 +711,7 @@
 - func: gesv(Tensor self, Tensor A) -> (Tensor, Tensor)
 
-- func: gesv_out(Tensor solution, Tensor lu, Tensor self, Tensor A) -> (Tensor, Tensor)
+- func: gesv_out(Tensor result0, Tensor result1, Tensor self, Tensor A) -> (Tensor, Tensor)
```
same here, please use the more informative names

animesht replied:
That's weird, I renamed these earlier and the codegen wouldn't allow me to pass `Tensor&` instead of `const Tensor&` if I didn't call them `result0, result1`, but I changed them back now and it seems to work...

SsnL commented on tools/autograd/derivatives.yaml:
```diff
@@ -307,7 +307,7 @@
 - name: _gesv_single(Tensor self, Tensor A)
   self: gesv_backward_self(grad, self, A)
-  A: gesv_backward_A(grad, self, A, solution)
+  A: gesv_backward_A(grad, self, A, result0)
```
changing to `result0` here is fine because it is now a native function

SsnL commented on aten/src/ATen/native/cuda/Gesv.cu:
```diff
@@ -138,6 +160,35 @@ std::tuple<Tensor,Tensor> _gesv_helper_cuda(const Tensor& self, const Tensor& A)
   return std::tuple<Tensor,Tensor>(b_working_copy, A_working_copy);
 }
 
+std::tuple<Tensor&,Tensor&> _gesv_single_out_cuda(Tensor& result0, Tensor& result1,
```
same here, let's use the more informative names

ssnl reviewed (commented):
Overall looks great! Some small nits.

zou3519 commented on aten/src/ATen/native/cuda/Gesv.cu:
```diff
@@ -138,6 +160,35 @@ std::tuple<Tensor,Tensor> _gesv_helper_cuda(const Tensor& self, const Tensor& A)
+  int info;
+  int* ipiv;
+
+  result1 = A.t().clone();
```
What is the `.t().clone()` for? Is this to ensure result1, result0 are in column major format?

If so, this code assumes that A, self are contiguous and they might not be

zou3519 replied:
edit: I thought about it some more and `.t().clone()` is fine.

You could call https://github.com/pytorch/pytorch/blob/71626491c40c86eb1ee410e695c467486084144c/aten/src/ATen/native/LinearAlgebraUtils.h#L15 instead (and probably rename the function to something like "cloneColumnMajor") so this becomes clearer.

zou3519 commented on aten/src/ATen/native/cuda/Gesv.cu:
```diff
@@ -138,6 +160,35 @@ std::tuple<Tensor,Tensor> _gesv_helper_cuda(const Tensor& self, const Tensor& A)
+    "compilation. Please rebuild with MAGMA.");
+#else
+  int64_t bx = self.size(0);
+  int64_t by = (self.dim() == 1) ? 1 : self.size(1);
```
Could you update the docs on this? The docs currently say that `gesv` only supports 2D self (B variable).

It looks like the old THCTensor_(gesv) doesn't support 1D self but the THTensor_(gesv) supports both 2D and 1D self:
https://github.com/pytorch/pytorch/blob/7b375ed362c4da833c38691b6bf3ade0941d3bf1/aten/src/THC/generic/THCTensorMathMagma.cu#L66-L67

zou3519 commented on aten/src/ATen/native/Gesv.h:
```diff
@@ -2,17 +2,32 @@
+          (long long)A.size(0), (long long)A.size(1));
+    } else if (A.size(0) != self.size(0)) {
+      AT_ERROR("A,B size incompatible - A has %ld "
+          "rows, B has %ld cols", A.size(0), self.size(0));
```
Should this be %lld as well?

zou3519 reviewed (commented):
This looks good, @animesht! I had a few minor comments and questions, in particular:
1) As @SsnL mentioned, it would be nice if we could rename (result0, result1) to (solution, lu) (or whatever the ordering is) if the code generation allows for it.
~~2) Tensors sent into the magma/lapack gesv bindings are required to be in column-major format: I'm not sure if the `.t().clone()` that is being done is sufficient for this~~ nvm, this is fine!

zou3519 commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +72,34 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+  int64_t by = (self.dim() == 1) ? 1 : self.size(1);
+  int info = 0;
+
+  result1 = A.t().clone();
```
C++ question: does this reassign the reference `result1` to the output of `A.t().clone()` or does this "copy" `A.t().clone()` into `result1`?

ezyang replied:
I discussed and this code is not right.

animesht replied:
I discussed this with @ezyang, it is the former -- I'll correct it in the next revision to check sizes and copy from the input matrices inplace

animesht committed:
Addressed comments

animesht commented:
- Renamed `result0, result`
- Fixed docs to mention that `B` can be 1D

animesht commented:
@pytorchbot retest this please

animesht committed:
Fixed copying for output Tensors

animesht committed:
Use contiguous Tensor while copying B

animesht committed:
Fixed view/contiguous stuff again...

animesht committed:
Make A contiguous before copy

animesht committed:
Handle Tensor reuse

animesht committed:
Finally fixed the last issue with torch.gesv(b, A, out=(b,A))

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -43,6 +43,12 @@ template<> void lapackGesv<double>(
 #endif
 
+static inline bool isTransposeContiguous(Tensor& self) {
+ return self.dim() == 2 &&
```
you can optimize this further by allowing dimension with size `<= 1` to have arbitrary stride

SsnL replied:
Here is what I meant: If a dimension has size `<= 1`, it can have arbitrary stride at that dimension and still be contiguous because the index at that dimension is at most `0`.

SsnL replied:
It's up to you whether to add more boolean statements :) This is fine as is by me.

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +78,65 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+  } else if (sol.numel() == self.numel() &&
+             isTransposeContiguous(sol)){
+    // allow reuse
+    sol.t_().copy_(self.view({bx, by}).t());
```
this breaks 
1. if `self` is not viewable to `{bx, by}`, or
2. `sol` doesn't have correct shape (but has correct `numel`)

animesht replied:
1. since `bx` and `by` are dimensions of `self`, it should always be viewable as `{bx, by}` (this is done to simplify the code for handling both 1D and 2D `b` Tensors.
2. Fixed in the latest commit to always call resize to the target output size regardless of contiguity or `numel`

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +78,65 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+  // init to column major format
+  // Note: `sol` and `lu` need to be contiguous since we pass
+  // sol.data() and lu.data() to Lapack
+  if (&self == &sol) {
```
As we discussed, this is not the comprehensive check. But this is fine as an easy optimization. Can you add a small comment here to explain?

Also, the code still works correctly if you remove this optimization, right?

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +78,65 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+    // allow reuse
+    sol.t_().copy_(self.view({bx, by}).t());
+  } else {
+    sol.resize_({by, bx}).copy_(self.view({bx, by}).t());
```
this also breaks if `self` is not viewable to `{bx, by}`.

animesht replied:
discussed

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +78,65 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+    }
+  } else if (lu.numel() == A.numel() &&
+             isTransposeContiguous(lu)) {
+    lu.t_().copy_(A.t());
```
same concerns here.

animesht replied:
fixed

ssnl reviewed (changes requested):
I reviewed cpu version. But the exact comments applies to the CUDA version.

animesht committed:
Addressed comments

animesht committed:
Use data_ptr() instead of address for comparison

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +78,63 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+  bool tc = sol.dim() == 2 && isTransposeContiguous(sol);
+  if (tc) {
+    // if transpose of output tensor is contiguous, use it
+    sol.t_();
```
I'm not understanding why this matters? Can you enlighten me? It looks like `sol` can be `resize_`d below anyways, right?

animesht replied:
<img width="511" alt="screen shot 2018-08-01 at 4 46 01 pm" src="https://user-images.githubusercontent.com/1777276/43554652-722cc238-95aa-11e8-93f7-01bd731a227c.png">
So in the test cases and the documentation the Tensors are initialized and then transposed before being passed to `torch.gesv`. We have to transpose it internally as well for Lapack/Magma's column major format and we need the data at`A_ptr` and `b_ptr` to be contiguous before we pass it. If a user calls `torch.gesv(b, A, out=(b, A))` and has transposed the input Tensor since creating it, we can just transpose `sol` here and use it directly.  The resize will not have any effect if it is already the required size (eg. the above call) but it will take care of Tensors whose transpose is contiguous but size is incorrect. Also, this helps with reusing output tensors;

for example:

    ta = torch.Tensor()
    tb = torch.Tensor()
    torch.gesv(b, a, out=(ta, tb)) # this will work
    torch.gesv(b, a, out=(ta, tb)) # this will not
The second time `gesv` is called, ta and tb will have been generated as a result of Lapack/Magma and transposed before being returned (our convention is to always return the transpose according to documentation). So even though they may have the correct size (resize_ won't work as expected on a non-contiguous Tensor), we should use their transpose if it is contiguous since we need pointers to contiguous data.

SsnL replied:
Thanks. But doesn't this still break if `sol` is not transpose contiguous but have size `{bx, by}`? I don't see any code handling the logic that in that case (1) applying lapack gesv on a buffer and then (2) copy to `sol`. Am I missing anything?

animesht commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +78,63 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+  }
+
+  sol.resize_({by, bx});
+  if (self.data_ptr() == sol.data_ptr()) {
```
If I remove this optimization and always do `sol.copy_(self.view({bx, by}).t().clone()` it will work. This code handles three cases:
1) When `sol` and `self` are the same Tensor (I'll revert to the previous address check)
    a) `sol` is transpose contiguous, so we transpose it (need to transpose input B anyway for Lapack)
    b) `sol`'s transpose is not contiguous, so it should be contiguous, but we need to copy the transpose of the input (`self`) into this contiguous tensor before passing it to Lapack. Since they share underlying storage, `self.view(...).t()` is cloned before copying to sol.
2) When `sol` and `self` are different tensors: we can do the copy to `sol` from `self` without the extra `.clone()`

SsnL replied:
Thanks. I think there might be one case that is not covered. See my other comment.

animesht committed:
Revert to address check, add missing case

SsnL commented:
@pytorchbot retest this please

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +78,100 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+
+  bool use_temp_sol = false;
+  bool use_temp_lu = false;
+  auto temp_sol = self.type().tensor();
```
Can you not allocate temporary tensors if not necessary?

animesht replied:
Switched to at::optional

animesht replied:
Since `at::optional` didn't work with MSVC, switched to nullable Tensors and using `t.defined()` to do the appropriate allocations

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +78,100 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+  bool tc = isTransposeContiguous(sol);
+  if (tc) {
+    // if transpose of output tensor is contiguous, use it
+    sol.t_();
```
I'm worried about this. It makes the function not re-entrant anymore. If an error is thrown before the final `sol_` near the end of this function, it is transposed forever! This can very well happen if input is half type or integral types because of the `AT_DISPATCH_FLOATING_TYPES` below. Can we not use the inplace version?

animesht replied:
Switched to using sol = sol.t(). This still reassigns the reference though

animesht replied:
moved the AT_ERROR out of AT_DISPATCH_FLOATING_TYPES to the end, after sol and lu have been transposed again. Should fix the re-entrant safeness

ssnl reviewed (commented):
Have two questions. Otherwise looks fine.

animesht committed:
Use at::optional, out-of-place transpose

animesht committed:
Fixed re-entrant safeness

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +80,101 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+    AT_ERROR("gesv only takes float or double Tensors");
+  }
+  checkErrors({info});
+  return std::tuple<Tensor&, Tensor&>(sol, lu);
```
You are not returning the original sol, but sol.t().t().

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +80,101 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+  } else if (sol.dim() == 2 &&
+             sol.size(0) == bx &&
+             sol.size(1) == by) {
+    temp_sol = self.type().tensor();
```
why not just `self.t().clone().t()`?

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +80,101 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+  }
+
+  if (res == -1) {
+    AT_ERROR("gesv only takes float or double Tensors");
```
This should already error at `AT_DISPATCH_FLOATING_TYPES`.... please revert this.

SsnL commented on aten/src/ATen/native/cuda/Gesv.cu:
```diff
@@ -138,6 +168,97 @@ std::tuple<Tensor,Tensor> _gesv_helper_cuda(const Tensor& self, const Tensor& A)
+  int64_t ay = A.size(1);
+  int info, res;
+  int* ipiv;
+  at::optional<Tensor> temp_sol;
```
msvc doesn't seem to like this...

animesht committed:
Revert error check, simplify temps, try to fix MSVC/windows build

animesht committed:
Merge branch 'master' of https://github.com/animesht/pytorch

animesht committed:
Fixes for MSVC and re-entrant safeness

animesht commented:
- I removed earlier transposes for the output tensors and moved the `checkErrors` call to immediately after the `AT_DISPATCH` block, so that any resulting errors are caught as early as possible. Until this point, `sol` and `lu` have not been modified (except the `resize_` if required), so it should be re-entrant safe. Before returning the final solution tensors, I transpose them in-place once. This should be fine since there is no way for the function to error out after this point. I switched from reassigning the reference `sol = sol.t()` to using the inplace transpose since the documentation mentions requirements for the output tensors:

> Irrespective of the original strides, the returned matrices X and LU will be transposed, i.e. with strides like B.contiguous().transpose(-1, -2).strides() and A.contiguous().transpose(-1, -2).strides() respectively.

- Since MSVC didn't seem to like `at::optional`, I switched to the following style:
```
Tensor temp_sol;
if (cond) { temp_sol = ... }
...
if (temp_sol.defined()) {...}

ezyang commented:
@animesht I reported the Windows problem upstream at https://github.com/akrzemi1/Optional/issues/58

Does the error message I reported there match the message you saw?

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +78,107 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+  /* Init to column major format. `sol` and `lu` need to be contiguous
+   * since we pass sol.data() and lu.data() to Lapack */
+  bool tc_sol = isTransposeContiguous(sol);
+  auto self_t = self.view({bx, by}).t();
```
I think this breaks if `self` is not viewable to `{bx, by}`.

SsnL replied:
There are only two real blockers:
1. This one.
2. Make the code clearer with restructuring / comment.

animesht replied:
`self` will always be viewable to `{bx, by}` since they are the dimensions of self. We discussed this earlier, I use `self.view({bx, by})` so that I don't explicitly have to unsqueeze in the case of 1D B (`self`) tensors.

SsnL replied:
Ah yeah.. you are right. My bad. This gets me every time. Can you add a comment here?

animesht replied:
Sure

ngimel replied:
Is self guaranteed to be contiguous? Otherwise it still cannot be viewed.

animesht replied:
I basically do this to treat 1D tensors as 2D by viewing them as `{t.size(0), 1}`. If self is 2D it is just the same as viewing it with it's dimensions `{t.size(0), t.size(1)}`.

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +78,107 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+
+  if (!temp_sol.defined()) {
+    if (tc_sol) {
+      sol.t().resize_({by, bx});
```
This deserves a comment.

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +78,107 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+        sol.copy_(self_t);
+      }
+    }
+  }
```
I'm very confused. At this point, should `sol` be transposed or not? They kind of make sense to me but I am still not sure if they are correct, because so many things are controlled by `tc_sol` being `true` or `false` that it is difficult to keep track.

I think it would be better to make the code clearer. It would be better for future maintenance too. Can you write these if statements so that they are
```
if (tc_sol) {
  // everything up till here...
} else {
  // everything up till here...
}
```
or write out your strategy clearly in a comment block. Thanks.

SsnL replied:
Additionally, I wonder if this kind of structure will make it clearer.

```c++
// gesv takes in two tensors and returns them in-place. so unless the output
// args are the same as input args, we will need to make a copy.

// these tensors will be the input and output args to the lapack call.
// they may be sol and lu, or may be a buffer, indicated by flags sol_buffer
// and lu_buffer.
Tensor sol_lapack_arg;
Tensor lu_lapack_arg;

bool sol_correct_shape = sol.size(0) == bx && sol.size(1) == by;
bool sol_buffer = false;

// now we detect if we need to use a buffer and/or resize the out arguments.
//
// our strategy is the same for each out arg. bellow we use (sol, self) pair
// for demonstration purpose. 
// 1. if sol has correct shape and is fortran contig
//    copy self into sol (if self is not sol)
//    call lapack gesv using sol
// 2. if sol has correct shape but isn't fortran contig
//    make a buffer, copy self into the buffer s.t. it is fortran contig
//    call lapack gesv using the buffer
//    copy the buffer content into sol
// 3. if sol has incorrect shape, then we have to resize
//    resize sol inplace, and make it fortran contig and holding content of self
//    call lapack gesv using sol

if (sol_correct_shape && tc_sol) {
    if (&sol != &self) {
        sol.copy_(self);
    }
    sol_lapack_arg = sol;
} else if (sol_correct_shape) {  // shape right but not fortran contiguous
    sol_lapack_arg = self.t().clone().t_(); // make it fortran contiguous
    sol_buffer = true;
} else {  // wrong shape, resize_
    // since shape is wrong, resize_ call will make it (fortran) contiguous
    sol_lapack_arg = sol.resize_({by, bx}).t_().copy_(self);  
}

// same for lu

AT_DISPATCH_FLOATING_TYPES(self.type(), "gesv", [&]{
    auto ipiv = at::empty({bx}, sol.options().dtype(kInt));
    lapackGesv<scalar_t>(bx, by, lu_lapack_arg.data<scalar_t>, bx, 
                         ipiv.data<int>(), sol_lapack_arg.data<scalar_t>, 
                         bx, &info);
    });
checkErrors({info});

if (sol_buffer) {
    sol.copy_(sol_lapack_arg);
}

// same for lu

return std::make_tuple(sol, lu);
```

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +78,107 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+  if (temp_sol.defined()) {
+    sol.copy_(temp_sol.t_());
+  } else if (!tc_sol) {
+    sol.t_();
```
Why do you need this inplace transform here? I think you should just copy into `sol` in a way so that it is column major contiguous already, right?

animesht replied:
If `tc_sol` or `tc_lu` is true, we're implicitly using their transposed version all along since it is contiguous and we need to give Lapack a pointer to contiguous data. After Lapack has been called, since the original sol/lu already had a transposing view on it, we can just return it. However if `tc_sol`/`tc_lu` is false, then we need to transpose the result from Lapack before returning it.

SsnL replied:
Yes, I agree with you that `t_` is needed with the current code. I was thinking if in the 3rd case you did `sol.resize_({by, bx}).copy_(self.t()).t_();`, the `t_` here isn't needed, and it makes the code easier to read since you have the invariant that no matter which path you take in the big `if` above, `sol` has correct shape. What do you think?

animesht replied:
Yeah I put it here since we wanted to avoid using t_() before all error checks were done, but I can move it to case 3

animesht committed:
Simplified logic, added notes

animesht commented:
@SsnL I incorporated some of your suggestions and added some notes at the top to make the structure more clear. Hope this looks better!

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +78,118 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+   * For 2D matrices, A.t() and self.t() represent their column major formats.
+   *
+   * Case 1) The output tensor is of the correct shape, but it and its transpose
+   *         are not contiguous. eg. torch.gesv(... , out=(n[::2], ...)):
```
Why does output tensor itself not being contiguous matter here?

animesht replied:
We should only create a buffer when we don't have a contiguous Tensor to work with, so if the output tensor is contiguous it should be handled properly in Case 3 without the need for a temp buffer

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +78,118 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+
+  /* self is always viewable to {bx, by} since they are the dimensions
+   * of self (or by == 1). Basically a shortcut to see 1D `self` as 2D */
+  auto self_t = self.view({bx, by}).t();
```
small nit: you can inplace t_ here.

animesht replied:
if used `.view().t()` since we may not actually need to use `self_t` in one case (when `tc_sol == true` and `&sol == &self`), so we can avoid doing additional work if it's not being cloned. Are there any other benefits to doing an inplace `t_()` here?

SsnL replied:
Ah makes sense! Thanks! So the above check is really testing "if the elements in `sol` form a contiguous chunk of array of size bx * by", right? Can you make this clearer in the comments above?

ssnl reviewed (commented):
Thanks! The comment helps greatly. I mostly have one question which I asked inline below.

animesht committed:
Addressed comments

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +78,114 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+  if (!tc_sol && !sol.is_contiguous() && sol_correct_shape) {
+    temp_sol = self_t.clone().t_();
+  } else if (tc_sol) {
+    sol.t().resize_({by, bx});
```
Can you move this into the if? It doesn't do anything if self = sol.

animesht replied:
fixed

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +78,114 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+  } else {
+    sol.resize_({by, bx});
+    if (&self == &sol) {
+      sol.copy_(self_t.clone()).t_();
```
It's worth a comment here: we can't use `self` here because `self` is `sol` and after `resize_` above it may just contain garbage.

animesht replied:
fixed

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +78,114 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+   * For 2D matrices, A.t() and self.t() represent their column major formats
+   *
+   * Case 1) The output tensor is of the correct shape, but it and its transpose
+   *         are not contiguous. eg. torch.gesv(... , out=(n[::2], ...)):
```
This is still a bit confusing because what you really want to test if the elements form a contiguous chunk of memory. I would instead say something along the lines of:
```
The output tensor is of the correct shape, but its elements do not form 
a contiguous chunk of memory (i.e., it and its transpose are not contiguous). 
Since shape is correct, we must not resize_ it, but clone input into a 
buffer, use the buffer for lapack, and copy to output afterwards.
```

animesht replied:
fixed

animesht committed:
Addressed comments

curry-sauce commented:
@animesht code is pretty unreadable mate

mazinbokhari commented:
really great work @animesht ! 😄

SsnL commented:
But before merging. @ezyang @Jorghi12 why is rocm build not skipping that gesv test (see the CI failure)? Is the USE_LAPACK flag not properly set in ATen with rocm build?

SsnL commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +78,111 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+    Tensor& sol, Tensor& lu,
+    const Tensor& self, const Tensor& A) {
+#ifndef USE_LAPACK
+  AT_ERROR("gesv: LAPACK library not found in compilation");
```
Ah okay I know why the build's failing. We had a hacky `skipIfLapack` rule that checks the error message but this isn't fitting the pattern. LOL

Can you make this throw `gesv : Lapack library not found in compile time` instead? Would be great to change the one above too. Thanks! :) I'll add a proper flag for LAPACK later.

animesht committed:
Fix error message

mazinbokhari commented:
Your mother and I are very proud of you chotu  @animesht

vishwakftw commented on aten/src/ATen/native/Gesv.h:
```diff
@@ -2,17 +2,31 @@
+  if (batched) {
+    if (A.size(-1) != A.size(-2)) {
+      AT_ERROR("A must be batches of square matrices, "
+          "but they are %lld by %lld matrices",
```
I don't think the formatting works. I faced an issue with this earlier.

animesht replied:
I wonder why they used this format specifier, lol, I remember seeing it didn't work but I thought it was because we use `setup.py build develop`. changed it anyway

vishwakftw commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -43,10 +43,16 @@ template<> void lapackGesv<double>(
+}
+
 template <typename scalar_t>
 static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
```
I think this was missed earlier, but `infos` has to be passed by reference to actually flag errors in `_gesv_helper_cpu`. This is not part of the PR, but I thought I should let you know.

vishwakftw replied:
I found this is https://github.com/pytorch/pytorch/pull/9949 , and this change might break some tests for `gesv`.

animesht replied:
I'm confused. Isn't `info` being created and passed by reference below on line 63? `infos` is just accumulating the results for each batch of matrices, and we return from `applyGesv` as soon as we see an error.

SsnL replied:
Yes @vishwakftw is correct. @animesht do you mind making this change? :)

animesht replied:
#10492

animesht committed:
renamed vars, pass infos by reference

animesht committed:
Fix formatting issues with %lld format specifier

animesht committed:
Revert pass by reference

SsnL commented:
@pytorchbot retest this please

zou3519 commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -43,10 +43,16 @@ template<> void lapackGesv<double>(
 #endif
 
+static inline bool isTransposeContiguous(Tensor& self) {
+ return self.dim() == 2 &&
```
Nit: spacing (return is indented by only 1 space right now)

zou3519 commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +78,111 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+   *            - clone and copy input_tensor.t() to output_tensor (same tensor)
+   *         b) &input_tensor != &output_tensor:
+   *            - copy input_tensor.t() to output_tensor
+   */
```
Why are there so many cases? The original `THTensor_gesv(...)`code is much simpler than this and doesn't have `.t()` and `.t_()` everywhere. Did you improve on the original implementation?

animesht replied:
Yeah, this optimizes for the cases mentioned in this comment, the old `THTensor_gesv` code simply cloned the input tensors using `cloneColumnMajor` in every case. This will save a few copies and construction of new Tensors   in `gesv_single_out` for certain cases.

animesht replied:
`cloneColumnMajor` and `checkTransposed` in the TH code do something similar, I can refactor this to make it a function that can be shared by both the CPU and CUDA implementations

zou3519 replied:
Understood. Yes, please do -- that would greatly improve the readability of the code as well

zou3519 commented on aten/src/ATen/native/cuda/Gesv.cu:
```diff
@@ -138,6 +166,76 @@ std::tuple<Tensor,Tensor> _gesv_helper_cuda(const Tensor& self, const Tensor& A)
+AT_ERROR("gesv: MAGMA library not found in "
+    "compilation. Please rebuild with MAGMA.");
+#else
+  /* See Gesv.cpp for comments */
```
A lot of this code is duplicated between gesv.cpp and gesv.cu. Could we refactor out the common bits?

As I understand it most of this code is coercing the inputs into column major format, right?

zou3519 reviewed (commented):
Haven't gone through the code in detail yet. I left some comments on refactoring the code to be a little nicer

animesht committed:
Refactored

animesht commented:
@zou3519, addressed your comments and refactored the implementation to `Gesv.h` and made it more compact. Hope this looks better! 

cc @SsnL

animesht committed:
Fixed typo

animesht committed:
Fix indent

animesht committed:
fix jenkins

animesht committed:
fix jenkins

yf225 commented:
@pytorchbot retest this please

zou3519 commented on aten/src/ATen/native/Gesv.h:
```diff
@@ -1,30 +1,92 @@
+ *    (i)  &in == &out: copy in.t().clone() to out (same tensor)
+ *    (ii) &in != &out: copy in.t() to out
+ */
+static inline void prepareIOTensors(
```
What does "prepareIOTensors" do? This comment gives a few cases but I'm not sure what this function is trying to accomplish (what is IO? input-output?) It would be great if this function name were more descriptive.

animesht replied:
how about `prepareTensorsForLapack`?

zou3519 replied:
That's better. What does it mean to prepare a tensor for lapack?

animesht replied:
basically, ensuring that pointers passed to Lapack represent Fortran contiguous tensors in column major format

zou3519 commented on aten/src/ATen/native/Gesv.h:
```diff
@@ -1,30 +1,92 @@
+ */
+static inline void prepareIOTensors(
+    const Tensor& in, Tensor& out, Tensor& temp,
+    int64_t& x, int64_t& y) {
```
So the function takes in `in` and returns `out`, `temp`, `x`, `y`. To my understanding, the output tensor always (sometimes?) has size (y, x): is it necessary to return both `x` and `y`?

animesht replied:
It's just easier to pass them by reference here so that I don't have to get (y, x) from the output tensor (which may not have .size(1), so I might have to do `dim == 1 ? 1 : .size(1)` again. I don't really use the values written for `ax, ay` (`A`'s dimensions), but pass `bx`, `by` to Lapack.

animesht replied:
Actually, I agree -- it'll probably be better to just get `bx`, `by` in the `AT_DISPATCH` block and use it instead of returning from the helper. Will change this in the next commit

zou3519 replied:
Does the output tensor always have size (y, x)? Sorry, I'm a little confused by the logic

animesht replied:
Yeah, the output tensor will be 2D and have size (y, x) after the prepare call. If `temp_sol` is being used, `sol` was already the correct shape, and in the case where it is 1D it will be resized to be 2D, so we can always do:
```
AT_DISPATCH_FLOATING_TYPES(...
    int64_t bx = sol.size(1);
    int64_t by = sol.size(0);
    ...
})

animesht committed:
Addressed comments

animesht committed:
Addressed comments

animesht committed:
fixed typo

zou3519 commented on aten/src/ATen/native/Gesv.h:
```diff
@@ -1,30 +1,94 @@
+ * (ii) for 2D matrices, .t_() represents their column-major format
+ *
+ * Before passing pointers to Lapack, we need to ensure that these pointers
+ * represent Fortran-contiguous tensors in column-major format
```
Thank you @animesht, this comment gives the code a lot more context and makes it easier to understand now :)

zou3519 commented on aten/src/ATen/native/Gesv.h:
```diff
@@ -1,30 +1,94 @@
+ * copy the buffer to the output tensor.
+ *
+ * 2) out.t() is contiguous:
+ *    (i)  &in == &out: use out.data() as is. Do nothing
```
This is a little overengineered; we don't guarantee that passing in the input tensor as the output tensor works for any other operation in pytorch. Since your code is correct that is nice to have, but I'm not sure a user will stumble upon this.

animesht replied:
The main reason for these optimizations is that the Lapack `gesv` interface takes input tensors which contain the outputs after the function is called, so it's nice to support that. Also, `test_gesv` explicitly checks for this case.

zou3519 commented on aten/src/ATen/native/Gesv.h:
```diff
@@ -1,30 +1,94 @@
+  // view potential 1D `in` as 2D
+  auto in_t = in.view({x, y}).t_();
+
+  if (!out_tc && !out.is_contiguous() && out_correct_shape) {
```
There are a lot of conditions mixed together here so it's still hard for me to tell what's going on in each case. i.e., in the `if (!out_tc)` case, what other booleans hold? Since out is being resized, I think that `out_correct_shape` is False, but I'm not entirely sure.

Could you change the structure of the if statements to something like the following (pseudocode)
```
if (out_correct_shape):
    if !out_tc and !out.is_contiguous():
       ...
else:
    if (out_tc):
        ...
    else:
        ...
```
I think that is what you're trying to express, please let me know if not otherwise

animesht replied:
Well, not exactly, since we still want to resize when it is not the correct shape, it's just that the `resize` in the latter two conditions doesn't do anything if `out_correct_shape` is true. I think this is probably the most succinct way to write it without making the if-else ladder long and unnecessarily nested (since the exact cases are described in the comment above, it shouldn't be too hard to understand)

zou3519 replied:
I'm having a hard time understanding why the cases are mutually exclusive (are they?). It'll probably be more clear if I wrote down a truth table, but at some point this exceeds my threshold for complexity...

zou3519 commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +72,40 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+  Tensor temp_sol;
+  Tensor temp_lu;
+  prepareTensorsForLapack(self, sol, temp_sol);
+  prepareTensorsForLapack(A, lu, temp_lu);
```
Would it be easier if `prepareTensorsForLapack` returned a Tensor& for the tensor that we're going to send into the `magmaGesv`? 

i.e.,
`prepareTensorForLapack(A, lu, temp_lu)` would always returns a reference. It could be a reference to lu, or a reference to a temp_lu because lu was not suitable. I'm trying to get rid of the ternary expression you have in the dispatch because that is a little confusing.

animesht replied:
Yeah, I was thinking along the same lines to remove the ternary operator, although we'll basically be doing a `temp.defined` in `prepareTensorsForLapack` instead of here. I'll make the change.

zou3519 commented on aten/src/ATen/native/Gesv.cpp:
```diff
@@ -72,6 +72,40 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+  prepareTensorsForLapack(A, lu, temp_lu);
+
+  AT_DISPATCH_FLOATING_TYPES(self.type(), "gesv", [&]{
+    const int64_t bx = sol.size(0);
```
nit: call these `n`, `nrhs` to match lapack notation?

yf225 commented:
ping @animesht :)

animesht committed:
Addressed comments

animesht commented:
Addressed most comments @zou3519 . Still need to add the tests for each case to `test_torch.py`, I actually had these as uncommitted python files in my old devgpu which I'm locked out of :(

Hoping for that to come back up soon, otherwise I can rewrite those and add them

animesht committed:
Added tests

animesht commented:
@zou3519 added tests for all cases. Can I get a codeowner review?

fmassa commented:
This got superseded by https://github.com/pytorch/pytorch/pull/6100
